### DISCUSSION TITLE
خطا در تائید عملیات روی درگاه بانک سامان

### DISCUSSION THREAD

**Discussion by gomnam1336** [Sina-Soltani/Parbad#377] [Q&A]
سلام من از پرباد برای درگاه سامان استفاده میکنم مشکل من اینه که وقتی تراکنش موفق میشه و میخواد برگرده ، روی سرور خطا میده و به علاوه از سایت لاگ اوت شده و در نتیجه تراکنش verify نمیشه، اما وقتی که بک میزنم میبینم که کاربر لاگ این هست. فکر میکنم مشکل از call back url باشه ، پروژه من روی .net core 3.1 هست

روی درگاه مجازی همه چی خوبه ولی روی درگاه سامان که میاد خطا میده و تراکنش verify نمیشه

```
var callbackUrl = Url.Action("VerifySuccess", "Panel", null, Request.Scheme);
var result = _onlinePayment.RequestAsync(invoice =>
           {
               invoice
                      .SetCallbackUrl(callbackUrl)
                      .SetAmount((Money)price*10)
                       .SetGateway("Saman")
                      //.UseParbadVirtual()
                      .UseAutoIncrementTrackingNumber();

               //if (viewModel.GenerateTrackingNumberAutomatically)
               //{
               //    invoice.UseAutoIncrementTrackingNumber();
               //}
               //else
               //{
               //    invoice.SetTrackingNumber(viewModel.TrackingNumber);
               //}
           }).Result;

           // Note: Save the result.TrackingNumber in your database.

           if (result.IsSucceed)
           {    model.paymentstatus = true;  
              // temp.SellCredit(id, orderid);
               return result.GatewayTransporter.TransportToGateway();
             
              
              // return View(model);
           }
           else
           {
               return View(model);
           }
```

  ```
public async Task<IActionResult> VerifySuccessAsync()
        {
            Successfulpayviewmodel model = new Successfulpayviewmodel();
            var invoice = await _onlinePayment.FetchAsync();

            // Check if the invoice is new or it's already processed before.
            if (invoice.Status != PaymentFetchResultStatus.ReadyForVerifying)
            {
                // You can also see if the invoice is already verified before.
                var isAlreadyVerified = invoice.IsAlreadyVerified;
                return Content("عملیات  پرداخت با شکست روبه رو شد لطفا بعدا اقدام فرمائید..");
            }

            // This is an example of cancelling an invoice when you think that the payment process must be stopped.
            //if (!Is_There_Still_Product_In_Shop(invoice.TrackingNumber))
            //{
            //    var cancelResult = await _onlinePayment.CancelAsync(invoice, cancellationReason: "Sorry, We have no more products to sell.");

            //    return View("CancelResult", cancelResult);
            //}

            var verifyResult = await _onlinePayment.VerifyAsync(invoice);
            if(!verifyResult.IsSucceed)
            {
                var result = await _onlinePayment.RefundCompletelyAsync(verifyResult.TrackingNumber);
            }
            if (verifyResult.IsSucceed && verifyResult.Status==PaymentVerifyResultStatus.Succeed)
            {
              
              
                model.paymentstatus = true;
                
            }
            // Note: Save the verifyResult.TransactionCode in your database.
            var tupleModel = new Tuple<IPaymentVerifyResult, Successfulpayviewmodel>(verifyResult, model);
            return View(tupleModel);
        }
```

**Comment by Sina-Soltani** [comment]
درود
چه نسخه ای رو نصب کردید؟

**Comment by gomnam1336** [comment]
Parbad.AspNetCore(1.3.0)
Parbad.Storage.EntityFrameworkCore(1.3.1)

**Reply by Sina-Soltani** [reply]
آخرین نسخه پکیج Parbad رو نصب کنید

**Reply by gomnam1336** [reply]
سلام آخرین نسخه رو نصب کردم ولی فرقی نکرد ، توی دیتابیس جدول transaction ,هر بار تراکنش دو سطر ایجاد میشه  یکی نوشته. Payment is successful  و دومی null زده، type هم یکی ۱ و دیگری ۰  هست. 
شاید علت. اینکه verifyنمیشه و وقتی برمیگرده کاربر رو log out نشون میده همینه.

**Comment by gomnam1336** [comment]
منظورتون خود پکیج parbad هست نه parbad .aspnetcore درسته؟

**Reply by Sina-Soltani** [reply]
پکیج هایی که نصب کردید به هر حال باید نصب باشن چون بهشون نیاز دارید.
اما پکیج اصلی Parbad شامل خود درگاه ها هست و وقتی آخرین نسخه منتشر میشه، یعنی اینکه آپدیتی یا اتفاقی داخلش افتاده که باید نصب بشه و در واقع ارتباطی با سایر پکیج ها نداره.
آخرین نسخه،‌ نسخه بتا هست. اما در مرحله تست توسط کاربران هست.
لطفا در صورت موفق عمل کردن اطلاع بدید تا از حالت بتا در بیارم و نسخه اصلی رو منتشر کنم.

**Comment by gomnam1336** [comment]
سلام آخرین نسخه رو نصب کردم ولی فرقی نکرد ، توی دیتابیس جدول transaction ,هر بار تراکنش دو سطر ایجاد میشه یکی نوشته. Payment is successful و دومی null زده، type هم یکی ۱ و دیگری ۰ هست.
شاید علت. اینکه verifyنمیشه و وقتی برمیگرده کاربر رو log out نشون میده همینه.

**Reply by Sina-Soltani** [reply]
درود.
من به دلیل نداشتن وقت کافی برای توسعه این پکیج، در حال حاضر کمک خاصی نمیتونم در این زمینه و همچنین رفع ایرادات موجود انجام بدم.

**Comment by gomnam1336** [comment]
چجوری میشه یه ستون به جدول payment با EFCore اضافه کرد؟ و به اون جدول داده اضافه کرد واز اون ستون داده رو خوند؟

**Reply by Sina-Soltani** [reply]
درود
روشی وجود نداره، چون پایگاه داده Parbad برای مصرف داخلی خود Parbad هست.
هر گونه اطلاعاتی که شما لازم داشته باشید، در تمام مراحل در شي Result به شما داده شده که میتونید در پایگاه داده برنامه خودتون ذخیره کنید.
